homeassistant/components/rfxtrx/__init__.py

```diff
@@ -245,6 +258,20 @@ def handle_receive(event):
         if device_id in device_events:
             hass.bus.fire(EVENT_RFXTRX_EVENT, event_data)
 
+        if config[CONF_AUTOMATIC_ADD]:
+            if device_id not in devices:
+                data = entry.data
+                data[CONF_DEVICES][event_data["data"]] = device_id
+
+                asyncio.run_coroutine_threadsafe(
+                    async_add_device(hass, entry, data), hass.loop
+                )
+
+                devices.add(device_id)
+
+    async def async_add_device(hass, entry: config_entries.ConfigEntry, data):
```

Comment:
We don't need to make this a coroutine function since we don't await inside it. It can just be a callback decorated with `@callback` imported from core.py.

Reply:
```suggestion
    @callback
    def async_add_device(hass, entry: config_entries.ConfigEntry, data):
```

Reply:
I've changed it, but doubt if it's ok like this. This function, including the async_update_entry cannot run in parallel for multiple events. Once data is copied, it should update the entry before a new call can run. Shouldn't I await this function and make it `async def` again?

```
@callback
    def async_add_device_to_config(hass, event, device_id):
        data = entry.data.copy()
        event_code = binascii.hexlify(event.data).decode("ASCII")
        data[CONF_DEVICES][event_code] = device_id
        hass.config_entries.async_update_entry(entry=entry, data=data)
```

Reply:
My comment below should solve this I think. If we're doing everything in one pass of the event loop, updating the entry data will be sequential for each event.